connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdScheduler.kt

```diff
@@ -134,4 +137,50 @@ internal class UplynkAdScheduler(
         adHandler.createAdBreak(it)
         adBreaks.add(UplynkAdBreakState(it, AdBreakState.NOT_PLAYED))
     }
-}
\ No newline at end of file
+
+    fun isPlayingAd(): Boolean {
+        return adBreaks.firstOrNull { it.state == AdBreakState.STARTED } != null
+    }
+
+    fun skipAd(ad: Ad) {
+        adHandler.onAdSkip(ad.customData as UplynkAd)
+    }
+
+    fun getUnWatchedAdBreakOffset(time: Duration): Duration? {
+        return adBreaks.firstOrNull { it.adBreak.timeOffset <= time && it.state == AdBreakState.NOT_PLAYED }?.adBreak?.timeOffset
+    }
+
+    fun getLastUnWatchedAdBreakOffset(time: Duration): Duration? {
+        return adBreaks.lastOrNull { it.adBreak.timeOffset <= time }?.let {
+            if (it.state == AdBreakState.NOT_PLAYED)
+                return it.adBreak.timeOffset
+            else return null
+        }
+    }
+
+    fun getAdBreakOffset(time: Duration): Duration? {
+        return adBreaks.firstOrNull { (time in it.adBreak.timeOffset..(it.adBreak.timeOffset + it.adBreak.duration)) }?.adBreak?.timeOffset
+    }
+
+    fun getAdBreakEndTime(time: Duration): Duration? {
+        return adBreaks.firstOrNull { (time in it.adBreak.timeOffset..(it.adBreak.timeOffset + it.adBreak.duration)) }?.adBreak?.let {
+            it.timeOffset + it.duration
+        }
+    }
+
+    fun getSkipToTime(currentAd: Ad, currentTime: Duration): Duration? {
+        val currentUplynkAd = currentAd.customData as UplynkAd
+        for (adBreak in adBreaks) {
+            for (ad in adBreak.ads) {
+                if (ad.ad == currentUplynkAd && ad.state == AdState.STARTED) {
+                    return if (currentTime >= adBreak.adBreak.timeOffset + currentAd.skipOffset.toDuration(DurationUnit.SECONDS)) {
```

Comment:
The `ad.ad == currentUplynkAd` should only ever match once, right? In that case, you could return sooner.

And again, `.seconds` is a thing. 😁 
```suggestion
                if (ad.ad == currentUplynkAd) {
                    return if (ad.state == AdState.STARTED && currentTime >= adBreak.adBreak.timeOffset + currentAd.skipOffset.seconds) {
```